PopulationMetadatafile.csv will contain population data per year for every iu

c2akula opened the pull request:
The postprocessing pipeline depends on a `PopulationMetadatafile.csv` file that contains IU-wise population numbers for each of the diseases. However, this data doesn't contain yearly population numbers, thus, assuming that an IU only contains a single, static population.

This PR adds the feature where the yearly population data can be added to each IU to the `PopulationMetadatafile.csv`. This will introduce a "Year" column containing the range of years for which there is population information for each of the IUs.

```csv
IU_CODE,Year,ADMIN0ISO3,Priority_Population_Oncho
AGO02063,1974,AGO,25774.23919
AGO02063,1975,AGO,26372.47
AGO02063,1976,AGO,27009.9274
AGO02063,1977,AGO,27691.89124
AGO02063,1978,AGO,28420.98812
AGO02063,1979,AGO,29201.25238
AGO02063,1980,AGO,30031.07005
AGO02063,1981,AGO,30917.70969
AGO02063,1982,AGO,31857.97218
AGO02063,1983,AGO,32849.41494
AGO02063,1984,AGO,33889.75891
AGO02063,1985,AGO,34982.75096
AGO02063,1986,AGO,36114.75003
AGO02063,1987,AGO,37295.09921
AGO02063,1988,AGO,38526.93482
AGO02063,1989,AGO,39818.516
AGO02063,1990,AGO,41170.88618
AGO02063,1991,AGO,42604.34007
AGO02063,1992,AGO,44109.30536
```

c2akula committed:
Implement creating dummy population file to make it easier to test IUData construction and population size calculations especially with longitudinal data

c2akula committed:
Compute prevalence when iu_data contains longitudinal population numbers

c2akula committed:
Get yearly population numbers for country and continent levels to compute total population count in each year

c2akula committed:
Add adapter function to the tests to convert from raw yearly population data to PopulationMetadatafile.csv format with yearly population numbers

adiramani commented on endgame_postprocessing/post_processing/composite_run.py:
```diff
@@ -8,36 +8,139 @@
-    ]
-    return np.array(populations)[:, np.newaxis, np.newaxis]
+    ]).reshape(len(ius), -1, 1)
+    # return np.array(populations)[:, np.newaxis, np.newaxis]
```
Nit: remember to clean this up

c2akula replied:
Fixed.

adiramani commented on endgame_postprocessing/post_processing/composite_run.py:
```diff
@@ -49,9 +152,12 @@ def build_composite_run(
             canonical_iu_runs[0][canonical_columns.COUNTRY_CODE].iloc[0]
         )
 
+    if type(total_population) is dict:
```
If I am understanding correctly, it will only be a dict if there are years in the population data. Does it make more sense to toggle it based on the `has_yearly_data` attribute of iu data instead?

c2akula replied:
It's not a very good abstraction at the moment and is temporary. The challenge with using `has_yearly_data` is that it doesn't say what the population information is structured as - scalar or dict or something else. By returning it as a dict, it is documented in the header.

There is a broader question as well. How much sense does it make for when the `IUData` class is constructed from the `PopulationMetadatafile.csv`, and if the latter doesn't contain a `Year`, should the `IUData` provide a uniform abstraction where the downstream code depending on it doesn't need to worry about the `yearly` population toggle. Otherwise, I'm wondering if it becomes a leaky abstraction. I agree the code as it stands is not neat, it's temporary and once the computations are correct, I'll go back to address this.

adiramani commented on endgame_postprocessing/post_processing/iu_data.py:
```diff
@@ -10,7 +12,7 @@
 
 def _is_valid_iu_code(iu_code):
-    return re.match(r"[A-Z]{3}\d{5}$", iu_code)
+    return re.match(r"[A-Z]{3}.{0,5}[\d]{5}$", iu_code)
```
Is the `.{0,5}` to account for the longer IU codes? Since those would be numbers should we do something like 

`re.match(r"^[A-Z]{3}\d{5,10}$", iu_code)`?

c2akula replied:
The `.` part of `.{0,5}` is a wildcard because I've seen the extended IU codes to contain an alphanumeric sequence between the country code and the `IU_ID` part.

adiramani commented on endgame_postprocessing/post_processing/composite_run.py:
```diff
@@ -8,36 +8,139 @@
-        canonical_iu_runs, iu_data
-    )
+    priority_populations = _get_priority_populations(canonical_iu_runs, iu_data)
+    iu_case_numbers = all_ius_draws * priority_populations
```
It is still possible for the input prevalence data and the population data to have a different number of years. It might be good idea to have a filter or check before here to make sure that the years in the all_ius_draws match the years in priority populations

c2akula replied:
Yeah, this is where our previous discussion of doing the filtration to ignore the out-of-range years comes in. Haven't gotten around to it.

c2akula committed:
Fix bug in computing the total number of IUs in country where in the presence of yearly population data, the calculation should use the number of unique IUs in the country, and not the total count. Because with yearly data, the IU_CODE will be duplicated.

c2akula committed:
Optimize IUData population lookups and fix iterator compatibility

- Improve performance of country and Africa level population aggregations
- Add comprehensive documentation to IUData class
- Fix numpy array compatibility with population iterators
- Ensure consistent behavior for both longitudinal and non-longitudinal data

c2akula committed:
Optimize IUData with consistent iterator API and enhanced documentation

- Refactor population data access methods to return iterators consistently
- Enhance _precompute_views for improved performance and reduced duplication
- Add comprehensive documentation for all population methods and properties
- Update composite_run to handle iterator-based population data
- Fix all tests to work with new iterator-based API

c2akula committed:
Add year range intersection and trimming for simulation data

Implement functionality to handle cases where simulation years don't match population metadata years by:
- Computing intersection of simulation and metadata year ranges
- Trimming simulation data to only include overlapping years
- Raising errors when no overlap exists
- Adding comprehensive tests for all trimming scenarios

c2akula committed:
Add test to check that the building composite runs from longitudinal data works correctly

adiramani commented:
I think the CI failing is just due to ruff failures, could you double check if they are easy to change or not?

c2akula committed:
Fix code formatting and linting issues

- Fix line length violations by wrapping long lines
- Remove trailing whitespace
- Fix type comparison issues
